pkg/server/sotw/v3/server.go

```diff
@@ -61,177 +72,78 @@ type server struct {
-	nonce     string
-	resources map[string]struct{}
+	// Local configuration flags for individual xDS implementations.
+	opts config.Opts
 }
 
-// process handles a bi-di stream request
-func (s *server) process(str stream.Stream, reqCh <-chan *discovery.DiscoveryRequest, defaultTypeURL string) error {
-	// increment stream count
-	streamID := atomic.AddInt64(&s.streamCount, 1)
-
-	// unique nonce generator for req-resp pairs per xDS stream; the server
-	// ignores stale nonces. nonce is only modified within send() function.
-	var streamNonce int64
-
-	streamState := stream.NewStreamState(false, map[string]string{})
-	lastDiscoveryResponses := map[string]lastDiscoveryResponse{}
-
-	// a collection of stack allocated watches per request type
-	watches := newWatches()
-
-	// node may only be set on the first discovery request
-	var node = &core.Node{}
-
-	defer func() {
-		watches.close()
-		if s.callbacks != nil {
-			s.callbacks.OnStreamClosed(streamID, node)
-		}
-	}()
-
-	// sends a response by serializing to protobuf Any
-	send := func(resp cache.Response) (string, error) {
-		if resp == nil {
-			return "", errors.New("missing response")
-		}
+// streamWrapper abstracts critical data passed around a stream for to be accessed
+// through various code paths in the xDS lifecycle. This comes in handy when dealing
+// with varying implementation types such as ordered vs unordered resource handling.
+type streamWrapper struct {
+	stream    stream.Stream // parent stream object
+	ID        int64         // stream ID in relation to total stream count
+	nonce     int64         // nonce per stream
+	watches   watches       // collection of stack allocated watchers per request type
+	callbacks Callbacks     // callbacks for performing actions through stream lifecycle
+
+	node *core.Node // registered xDS client
+
+	// The below fields are used for tracking resource
+	// cache state and should be maintained per stream.
+	streamState            stream.StreamState
+	lastDiscoveryResponses map[string]lastDiscoveryResponse
+}
 
-		out, err := resp.GetDiscoveryResponse()
-		if err != nil {
-			return "", err
-		}
+// Send packages the necessary resources before sending on the gRPC stream,
+// and sets the current state of the world.
+func (s *streamWrapper) send(resp cache.Response) (string, error) {
+	if resp == nil {
+		return "", errors.New("missing response")
+	}
 
-		// increment nonce
-		streamNonce = streamNonce + 1
-		out.Nonce = strconv.FormatInt(streamNonce, 10)
+	out, err := resp.GetDiscoveryResponse()
+	if err != nil {
+		return "", err
+	}
 
-		lastResponse := lastDiscoveryResponse{
-			nonce:     out.Nonce,
-			resources: make(map[string]struct{}),
-		}
-		for _, r := range resp.GetRequest().ResourceNames {
-			lastResponse.resources[r] = struct{}{}
-		}
-		lastDiscoveryResponses[resp.GetRequest().TypeUrl] = lastResponse
+	// increment nonce and convert it to base10
+	out.Nonce = strconv.FormatInt(atomic.AddInt64(&s.nonce, 1), 10)
```

Comment:
Minor: do we need this atomic increase here? The rest of this method is not synchronized anyway

Reply:
I think we do need this since N number of "sends" can increment the nonce counter which is shared. I could be wrong though I ported that from the previous code

Reply:
My understanding is that all the "sends" run in the same goroutine on a stream, so I do not expect the need to synchronize, but might be wrong